### PR TITLE
Dockerize downloader_app

### LINKED ISSUE #62: Containerized downloader_app
- Containerized downloader_app is part of issue #50 .
- The sequence of PR #57 .
- Create a structure for dockerfiles.
- Pass the tests into the container.
- Run the application in celery from crontab.

### PR DESCRIPTION
**Pull request by esloch**
- Resolves #62 .

### CODE REVIEW

File: downloader_app/ci/create_credentials.py [hunk 1]
```diff
@@ -8,19 +8,17 @@
 
 load_dotenv()
 
-
 DOWNLADER_PATH = os.path.join(BASE_DIR, 'downloader_app')
-
+CREADENTIALS_PATH = os.path.join(DOWNLADER_PATH, "credentials.json")
```
Comment (fccoelho) [inline downloader_app/ci/create_credentials.py:12]: CREDENTIALS_PATH

Comment (fccoelho) [inline downloader_app/ci/create_credentials.py:12]: CREDENTIALS_PATH

**Review by fccoelho — APPROVED** [review]

**Comment by esloch** [comment]
Follow-up, issue #60 :
  - Create functions for variable dates (argparser).
  - Generate the data to be inserted in the table Clima_Satelite
  - Move the shapes_region directory to Storage.
  - Add tasks in run_celery_downloader.

File: docker/crawlclima/start_cron_celery.sh [hunk 2]
```diff
@@ -1,4 +1,5 @@
 #!/usr/bin/env bash
+
 python docker/crawlclima/setup.py develop &
```
Comment (xmnlab) [inline docker/crawlclima/start_cron_celery.sh:3]: I think you don't need this `&` at the end of line 2 (https://mywiki.wooledge.org/BashSheet#:~:text=An%20ampersand%20does%20the%20same,for%20the%20former%20to%20end.)

File: docker/satellite/start_cron_celery.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+python docker/satellite/setup.py develop &
```
Comment (xmnlab) [inline docker/satellite/start_cron_celery.sh:3]: same here about the `&` character

File: downloader_app/ci/create_credentials.py (hunk 1 above)
Comment (xmnlab) [reply downloader_app/ci/create_credentials.py:16]: as @fccoelho commented there is still a typo here

Comment (xmnlab) [reply downloader_app/ci/create_credentials.py:16]: as @fccoelho commented there is still a typo here too

File: downloader_app/run_celery_downloader.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+WORK_DIR = os.path.join(BASE_DIR, "downloader_app")
+sys.path.insert(0, WORK_DIR)
+
+
+# Set the filepath and load in a shapefile.
+SHP_PATH = os.path.join(
+    WORK_DIR, "shape_region/RJ_Mun97_region/RJ_Mun97_region.shp"
```
Comment (xmnlab) [inline downloader_app/run_celery_downloader.py:20]: it shouldn't be hard-coded ... if you have plans to do it in another PR, add a `TODO` comment here.

File: downloader_app/run_celery_downloader.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+)
+DOWNLOADFILES_PATH = os.path.join(WORK_DIR, "DownloadedFiles")
+
+
+try:
```
Comment (xmnlab) [inline downloader_app/run_celery_downloader.py:25]: it looks like a test ... probably it should be removed from here

File: downloader_app/run_celery_downloader.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+if __name__ == "__main__":
+    pass
+
+
+class DownloaderSources:
+    def LandDAAC_v5_day():
+        # Parameters
+        # Extract bounding box from shapefile by uf
+        point1, point2 = shpm.extract_shp_boundingbox(SHP_PATH)
```
Comment (xmnlab) [inline downloader_app/run_celery_downloader.py:39]: probably SHP_PATH should be passed as a function argument ... not as a global variable

File: downloader_app/run_celery_downloader.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            'close_browser': True,
+        }
+
+        # Call taks
+        download_source.delay(source, dates, point1, point2, options)
+
+    def LandDAAC_v5_night():
+        # Parameters
+        # Extract bounding box from shapefile by uf
+        point1, point2 = shpm.extract_shp_boundingbox(SHP_PATH)
```
Comment (xmnlab) [inline downloader_app/run_celery_downloader.py:63]: same here about SHP_PATH

File: downloader_app/run_celery_downloader.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        # iterate source list
+        source = 'LandDAAC-v5-night'
+        # date_start date_end last week
+        # argparser(dt_start, dt_end)
+        dates = (
```
Comment (xmnlab) [inline downloader_app/run_celery_downloader.py:68]: it shouldn't be hard-coded here.
if you are planning to fix that in a follow-up PR, use the hard-coded parameters outside the functions and pass that as a parameter to the `DownloaderSources.LandDAAC_v5_day` call. ex:

```python
if __name__ == "__main__":
    dates = (
        pd.date_range('2016-07-20', '2016-08-05', freq='8D')
            .strftime("%Y-%m-%d")
            .tolist()
    )
    DownloaderSources.LandDAAC_v5_day(SHP_PATH, dates)
```

also, I think you will need to call other functions too like `LandDAAC_v5_night`, is it right? if yes, add a `TODO` comment that you will add it in a follow-up PR.

File: downloader_app/run_celery_downloader.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            pd.date_range('2016-07-20', '2016-08-05', freq='8D')
+            .strftime("%Y-%m-%d")
+            .tolist()
+        )
+        print(dates)
```
Comment (xmnlab) [inline downloader_app/run_celery_downloader.py:73]: remove this print, if you want to have this kind of option for debugging propose, you should use via a verbose flag. eg:

```python

def LandDAAC_v5_day(self, param1, param2, bla, verbose=False):
    some_code
    some_code_bla

    if verbose:
         print('bla')
```

but the verbose implementation is not required, you can just remove the print command

File: downloader_app/tasks.py [hunk 10]
```diff
@@ -4,16 +4,22 @@
...
-# sys.path.insert(0, LOCAL)
 
 load_dotenv()
 
-logger = get_task_logger("Downloader_app")
+logger = get_task_logger("downloader_app")
+
+
+@app.task
+def add(x, y):
```
Comment (xmnlab) [inline downloader_app/tasks.py:21]: probably you don't need the `add` function

File: downloader_app/tasks.py [hunk 11]
```diff
@@ -23,12 +29,12 @@ def download_source(source, dates, point1, point2, opt=False):
...
     try:
-        logger.info("Capturando {} {}".format(source, dates))
-        td.download_tiffs(source, dates, point1, point2, opt)
+        logger.info("Fetch {} {}".format(source, dates))
+        td(source, dates, point1, point2, opt)
 
     except Exception as e:
         logger.error(
-            "Error capturando from {} at {} error: {}".format(source, dates, e)
+            "Error fetch from {} at {} error: {}".format(source, dates, e)
```
Comment (xmnlab) [inline downloader_app/tasks.py:37]: you can simplify it using the standard EE for error and II for information, example:

`"[EE] fetching from {} at {} error: {}`

File: downloader_app/tests/test_tiff_downloader.py [hunk 12]
```diff
@@ -2,20 +2,24 @@
...
+)
+WORK_DIR = os.path.join(BASE_DIR, "downloader_app")
+sys.path.insert(0, WORK_DIR)
+
 
 # Set the filepath and load in a shapefile.
-SHP_PATH = os.path.join(LOCAL, "RJ_Mun97_region/RJ_Mun97_region.shp")
-DOWNLOADFILES_PATH = os.path.join(LOCAL, "DownloadedFiles")
+SHP_PATH = os.path.join(
+    WORK_DIR, "shape_region/RJ_Mun97_region/RJ_Mun97_region.shp"
```
Comment (xmnlab) [inline downloader_app/tests/test_tiff_downloader.py:20]: if you need a shapefile for test ... add it into a test folder, for example: `myproject_path/myapp_path/tests/data/RJ_Mun97_region.shp`

File: downloader_app/tiff_downloader.py [hunk 13]
```diff
@@ -62,26 +68,30 @@
 
 
 def download_tiffs(source, dates, point1, point2, opt=False):
-    """ 
-    Function responsible for the main calls. 
-    
+    """
+    Function responsible for the main calls.
```
Comment (xmnlab) [inline downloader_app/tiff_downloader.py:72]: for the first line of the docstring ... write it in an imperative form, for example:

`Download tiff files according to the given parameters.`

File: downloader_app/tiff_downloader.py [hunk 14]
```diff
@@ -62,26 +68,30 @@
...
     Inputs
     ------
     source: string
-        It is the name of the source from which we will take the captured images. 
+        It is the name of the source from which we will take the captured images.
         For example 'LandDAAC-v5-day' is a valid name. Other possible names may
         be seen with the command about('sources').
     dates: collection of timestamps
         The days of interest to download data.
     point1, point2: tuple
```
Comment (xmnlab) [inline downloader_app/tiff_downloader.py:81]: for tuple, list, etc you can be more specific ... for example tuple(string), list[str, int] etc. you can use the same approach used by mypy

File: setup.cfg [hunk 15]
```diff
@@ -4,11 +4,10 @@ license-file = LICENSE
 [flake8]
 max-line-length = 79
 ignore = D202,D203,W503,E203, E501, F841, E731
-exclude =
-    downloader_app/*
+exclude =""
```
Comment (xmnlab) [inline setup.cfg:7]: as you don't need this parameter anymore, you can just remove it

**Review by xmnlab — CHANGES_REQUESTED** [review]
thanks for working on this @esloch . I added some comments here ;)

File: downloader_app/tests/test_tiff_downloader.py (hunk 12 above)
Comment (esloch) [reply downloader_app/tests/test_tiff_downloader.py:20]: It's done!

**Review by esloch — COMMENTED** [review]
Thanks @xmnlab !

**Comment by xmnlab** [comment]
it seems you will need to fix some tests https://github.com/AlertaDengue/AlertaDengueCaptura/pull/63/checks?check_run_id=1179198658#step:9:58

File: downloader_app/run_celery_downloader.py (hunk 4 above)
Comment (fccoelho) [reply downloader_app/run_celery_downloader.py]: I agree. By the way, we should not be using any external shapefiles since we have the geojson for every municipality in the database.

File: downloader_app/tests/test_tiff_downloader.py (hunk 12 above)
Comment (fccoelho) [reply downloader_app/tests/test_tiff_downloader.py]: My Comment from above applies here as well,  do not use shapefiles. Use the Geojson from the database.

**Comment by fccoelho** [comment]
I commented on top of @xmnlab comments, so I may have missed some of the places where you use shapefiles. Please replace all shapefiles by the geojsons loaded from the database.

**Comment by esloch** [comment]
Thanks @fccoelho :)
It is a good idea to use geojson from the database, we have the  dbdemo in the CI of AD for tests that we can reuse here, but it would also be another PR... The code of the current PR reads only a single shapefile from (RJ), located at `tests/data`, this should be pointed by an environment variable, and for production/staging, it should point to a local folder.
The functionality of reading more than one shapefile will be implemented in a follow-up PR (related to issue #60 ), that aims to reuse the shapefiles already stored locally (for tests it would be located at `/test/data`). if we want or need, we can also recreate the functionality (from AlertaDengue) to convert geojsons to shapefile.

**Comment by fccoelho** [comment]
Ok about keeping this one "shapefile" for testing. But consider converting it to the geopackage  format which is a single file format.
Geopandas can read and save both.
